### ISSUE TITLE
2015.5.3: Batch mode runs incorrectly

### ISSUE THREAD

**Issue by nicholascapo** [saltstack/salt#25937]
Using batch mode for an operation starts with the correct number of minions per batch but quickly degrades to running one at a time.
In other words `-b N` quickly becomes `-b 1` in the middle of a salt run.
This seems related to (but different than) https://github.com/saltstack/salt/issues/24875, as we were seeing this behavior leading up to the the crash in that bug.

Also the output is frequently messed up and incomplete.

--------------------------------------------------------------------------------------------

```
$ salt --versions-report
                  Salt: 2015.5.3
                Python: 2.7.3 (default, Mar 13 2014, 11:03:55)
                Jinja2: 2.6
              M2Crypto: 0.21.1
        msgpack-python: 0.1.10
          msgpack-pure: Not Installed
              pycrypto: 2.6
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 13.1.0
                  RAET: Not Installed
                   ZMQ: 3.2.3
                  Mako: 0.7.0
               Tornado: 4.0.2
 Debian source package: 2015.5.3+ds-1~bpo70+2
```

```
$ salt ids* test.version
ids999:
    2015.5.3
ids5:
    2015.5.3
ids6:
    2015.5.3
ids4:
    2015.5.3
ids3:
    2015.5.3
ids2:
    2015.5.3
ids1:
    2015.5.3
```

```
$ salt ids* -b 2 test.ping

Executing run on ['ids999', 'ids3'] <====================== two

ids3:
    True
retcode:
    0
ids999:
    True
retcode:
    0

Executing run on ['ids2', 'ids5'] <====================== two

ids2:
    True
retcode:
    0

Executing run on ['ids4'] <====================== one!

ids5:
    True
retcode:
    0

Executing run on ['ids6'] <====================== one!

ids4:
    True
retcode:
    0

Executing run on ['ids1'] <====================== one!

ids6:
    True
retcode:
    0
ids1:
    True
retcode:
    0
```

Strangely this seems to not always happen, it never fails this test case:

```
$ salt ids* -b 2 cmd.run 'sleep 5s'

Executing run on ['ids1', 'ids3']

ids3:
retcode:
    0
ids1:
retcode:
    0

Executing run on ['ids2', 'ids5']

ids5:
retcode:
    0
ids2:
retcode:
    0

Executing run on ['ids4', 'ids6']

ids6:
retcode:
    0
ids4:
retcode:
    0

Executing run on ['ids999']

ids999:
retcode:
    0

```

But highstate runs are very strange:
```
$ salt ids* -b 2 state.highstate

Executing run on ['ids1', 'ids3']

ids3:
  Name: /etc/nginx/conf - Function: file.recurse - Result: Changed
  Name: nginx - Function: service.running - Result: Changed
  Name: /usr/local/v/deployer - Function: file.recurse - Result: Changed
  Name: v-deployer - Function: service.running - Result: Changed
  Name: /etc/elasticsearch/elasticsearch.yml - Function: file.managed - Result: Changed
  Name: elasticsearch - Function: service.running - Result: Changed
  Name: /etc/sensu/plugins/mysql-metrics - Function: file.managed - Result: Changed

Summary
--------------
Succeeded: 661 (changed=7)
Failed:      0
--------------
Total states run:     661

Executing run on ['ids2']

              ----------
              ids1:
                  ----------

Executing run on ['ids5']

              ----------
              ids2:
                  ----------

Executing run on ['ids4']

ids5:
  Name: /etc/nginx/conf - Function: file.recurse - Result: Changed
  Name: nginx - Function: service.running - Result: Changed
  Name: /usr/local/v/deployer - Function: file.recurse - Result: Changed
  Name: v-deployer - Function: service.running - Result: Changed
  Name: /etc/elasticsearch/elasticsearch.yml - Function: file.managed - Result: Changed
  Name: elasticsearch - Function: service.running - Result: Changed
  Name: /etc/sensu/plugins/mysql-metrics - Function: file.managed - Result: Changed

Summary
--------------
Succeeded: 661 (changed=7)
Failed:      0
--------------
Total states run:     661

Executing run on ['ids6']

              ----------
              ids4:
                  ----------

Executing run on ['ids999']

              ----------
              ids6:
                  ----------
ids999:
  Name: /etc/nginx/conf - Function: file.recurse - Result: Changed
  Name: nginx - Function: service.running - Result: Changed
  Name: /usr/local/v/deployer - Function: file.recurse - Result: Changed
  Name: v-deployer - Function: service.running - Result: Changed
  Name: /etc/elasticsearch/elasticsearch.yml - Function: file.managed - Result: Changed
  Name: elasticsearch - Function: service.running - Result: Changed
  Name: /etc/sensu/plugins/mysql-metrics - Function: file.managed - Result: Changed

Summary
--------------
Succeeded: 661 (changed=7)
Failed:      0
--------------
Total states run:     661

```

**Comment by jaybocc2** [comment]
also running 2015.5.3 and getting this output after 2 runs:

```
admin@saltshaker-001:~$ sudo amp_salt_deploy G@ec2_tags:cluster:dash '{"app_deploy": true}' 2

Executing run on ['dev-dash-002.us-west-2', 'prod-dash-005.us-west-2']

prod-dash-005.us-west-2:
  Name: core_packages - Function: pkg.installed - Result: Clean
  Name: deb https://get.docker.com/ubuntu docker main - Function: pkgrepo.managed - Result: Clean
  Name: lxc-docker - Function: pkg.latest - Result: Clean
  Name: /etc/default/docker - Function: file.managed - Result: Clean
  Name: docker - Function: service.running - Result: Clean
  Name: pip - Function: pip.installed - Result: Changed
  Name: requests - Function: pip.installed - Result: Changed
  Name: docker-py ==1.2.3 - Function: pip.installed - Result: Clean
  Name: ~admin/.ssh - Function: file.directory - Result: Clean
  Name: ~admin/.ssh/vault_id_rsa - Function: file.managed - Result: Clean
  Name: git@github.com:amplitude/vault.git - Function: git.latest - Result: Clean
  Name: /mnt/tmp/logs - Function: file.directory - Result: Clean
  Name: /mnt/tmp/logs/dash - Function: file.directory - Result: Clean
  Name: amplitudemobile/dash - Function: docker.pulled - Result: Clean
  Name: nginx - Function: docker.pulled - Result: Clean
  Name: /usr/local/bin/docker_compare_running - Function: file.managed - Result: Clean
  Name: /usr/bin/docker stop -t 10 dash && /usr/bin/docker rm -f dash - Function: cmd.run - Result: Clean
  Name: /usr/bin/docker stop -t 10 nginx && /usr/bin/docker rm -f nginx - Function: cmd.run - Result: Changed
  Name: dash - Function: docker.installed - Result: Clean
  Name: nginx - Function: docker.installed - Result: Changed
  Name: dash - Function: docker.running - Result: Clean
  Name: nginx - Function: docker.running - Result: Changed
  Name: /mnt/tmp/logs/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx/conf.d - Function: file.directory - Result: Clean
  Name: /mnt/nginx/static - Function: file.directory - Result: Clean
  Name: /bin/rm -rf /mnt/nginx/static/public  && /usr/bin/docker cp dash:/app/amplitude/dash/public /mnt/nginx/static - Function: cmd.wait - Result: Clean
  Name: /mnt/nginx/conf.d/amplitude.conf - Function: file.managed - Result: Clean
  Name: /mnt/nginx/proxy_params/proxy_params - Function: file.managed - Result: Clean
  Name: /mnt/nginx/conf.d/blockips.conf - Function: file.managed - Result: Clean
  Name: /usr/bin/docker kill -s HUP nginx - Function: cmd.wait - Result: Clean
  Name: /etc/hosts - Function: file.managed - Result: Clean
  Name: /etc/security/limits.d/001-core-limits.conf - Function: file.managed - Result: Clean
  Name: /mnt - Function: mount.mounted - Result: Clean
  Name: ~admin/.ssh/authorized_keys - Function: file.managed - Result: Clean
  Name: deb http://apt.datadoghq.com/ stable main - Function: pkgrepo.managed - Result: Clean
  Name: datadog-agent - Function: pkg.latest - Result: Clean
  Name: /etc/dd-agent/datadog.conf - Function: file.managed - Result: Clean
  Name: datadog-agent - Function: service.running - Result: Clean
  Name: /usr/local/bin/dogcat - Function: file.managed - Result: Clean
  Name: /bin/systemctl daemon-reload - Function: cmd.wait - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh - Function: file.managed - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh -k "$lic" -x "$key" -y "$secret" -v "$ver" - Function: cmd.run - Result: Changed
  Name: /mnt/tmp/docker - Function: file.directory - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: file.managed - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: cmd.run - Result: Changed

Summary
-------------
Succeeded: 46 (changed=7)
Failed:     0
-------------
Total states run:     46

Executing run on ['prod-dash-006.us-west-2']

              ----------
              dev-dash-002.us-west-2:
                  ----------

Executing run on ['stag-dash-002.us-west-2']

prod-dash-006.us-west-2:
  Name: core_packages - Function: pkg.installed - Result: Clean
  Name: deb https://get.docker.com/ubuntu docker main - Function: pkgrepo.managed - Result: Clean
  Name: lxc-docker - Function: pkg.latest - Result: Clean
  Name: /etc/default/docker - Function: file.managed - Result: Clean
  Name: docker - Function: service.running - Result: Clean
  Name: pip - Function: pip.installed - Result: Changed
  Name: requests - Function: pip.installed - Result: Changed
  Name: docker-py ==1.2.3 - Function: pip.installed - Result: Clean
  Name: ~admin/.ssh - Function: file.directory - Result: Clean
  Name: ~admin/.ssh/vault_id_rsa - Function: file.managed - Result: Clean
  Name: git@github.com:amplitude/vault.git - Function: git.latest - Result: Clean
  Name: /mnt/tmp/logs - Function: file.directory - Result: Clean
  Name: /mnt/tmp/logs/dash - Function: file.directory - Result: Clean
  Name: amplitudemobile/dash - Function: docker.pulled - Result: Clean
  Name: nginx - Function: docker.pulled - Result: Clean
  Name: /usr/local/bin/docker_compare_running - Function: file.managed - Result: Clean
  Name: /usr/bin/docker stop -t 10 dash && /usr/bin/docker rm -f dash - Function: cmd.run - Result: Clean
  Name: /usr/bin/docker stop -t 10 nginx && /usr/bin/docker rm -f nginx - Function: cmd.run - Result: Changed
  Name: dash - Function: docker.installed - Result: Clean
  Name: nginx - Function: docker.installed - Result: Changed
  Name: dash - Function: docker.running - Result: Clean
  Name: nginx - Function: docker.running - Result: Changed
  Name: /mnt/tmp/logs/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx/conf.d - Function: file.directory - Result: Clean
  Name: /mnt/nginx/static - Function: file.directory - Result: Clean
  Name: /bin/rm -rf /mnt/nginx/static/public  && /usr/bin/docker cp dash:/app/amplitude/dash/public /mnt/nginx/static - Function: cmd.wait - Result: Clean
  Name: /mnt/nginx/conf.d/amplitude.conf - Function: file.managed - Result: Clean
  Name: /mnt/nginx/proxy_params/proxy_params - Function: file.managed - Result: Clean
  Name: /mnt/nginx/conf.d/blockips.conf - Function: file.managed - Result: Clean
  Name: /usr/bin/docker kill -s HUP nginx - Function: cmd.wait - Result: Clean
  Name: /etc/hosts - Function: file.managed - Result: Clean
  Name: /etc/security/limits.d/001-core-limits.conf - Function: file.managed - Result: Clean
  Name: /mnt - Function: mount.mounted - Result: Clean
  Name: ~admin/.ssh/authorized_keys - Function: file.managed - Result: Clean
  Name: deb http://apt.datadoghq.com/ stable main - Function: pkgrepo.managed - Result: Clean
  Name: datadog-agent - Function: pkg.latest - Result: Clean
  Name: /etc/dd-agent/datadog.conf - Function: file.managed - Result: Clean
  Name: datadog-agent - Function: service.running - Result: Clean
  Name: /usr/local/bin/dogcat - Function: file.managed - Result: Clean
  Name: /bin/systemctl daemon-reload - Function: cmd.wait - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh - Function: file.managed - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh -k "$lic" -x "$key" -y "$secret" -v "$ver" - Function: cmd.run - Result: Changed
  Name: /mnt/tmp/docker - Function: file.directory - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: file.managed - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: cmd.run - Result: Changed

Summary
-------------
Succeeded: 46 (changed=7)
Failed:     0
-------------
Total states run:     46

Executing run on ['prod-dash-003.us-west-2']

              ----------
              stag-dash-002.us-west-2:
                  ----------

Executing run on ['prod-dash-004.us-west-2']

              ----------
              prod-dash-003.us-west-2:
                  ----------
prod-dash-004.us-west-2:
  Name: core_packages - Function: pkg.installed - Result: Clean
  Name: deb https://get.docker.com/ubuntu docker main - Function: pkgrepo.managed - Result: Clean
  Name: lxc-docker - Function: pkg.latest - Result: Clean
  Name: /etc/default/docker - Function: file.managed - Result: Clean
  Name: docker - Function: service.running - Result: Clean
  Name: pip - Function: pip.installed - Result: Changed
  Name: requests - Function: pip.installed - Result: Changed
  Name: docker-py ==1.2.3 - Function: pip.installed - Result: Clean
  Name: ~admin/.ssh - Function: file.directory - Result: Clean
  Name: ~admin/.ssh/vault_id_rsa - Function: file.managed - Result: Clean
  Name: git@github.com:amplitude/vault.git - Function: git.latest - Result: Clean
  Name: /mnt/tmp/logs - Function: file.directory - Result: Clean
  Name: /mnt/tmp/logs/dash - Function: file.directory - Result: Clean
  Name: amplitudemobile/dash - Function: docker.pulled - Result: Clean
  Name: nginx - Function: docker.pulled - Result: Clean
  Name: /usr/local/bin/docker_compare_running - Function: file.managed - Result: Clean
  Name: /usr/bin/docker stop -t 10 dash && /usr/bin/docker rm -f dash - Function: cmd.run - Result: Clean
  Name: /usr/bin/docker stop -t 10 nginx && /usr/bin/docker rm -f nginx - Function: cmd.run - Result: Changed
  Name: dash - Function: docker.installed - Result: Clean
  Name: nginx - Function: docker.installed - Result: Changed
  Name: dash - Function: docker.running - Result: Clean
  Name: nginx - Function: docker.running - Result: Changed
  Name: /mnt/tmp/logs/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx/conf.d - Function: file.directory - Result: Clean
  Name: /mnt/nginx/static - Function: file.directory - Result: Clean
  Name: /bin/rm -rf /mnt/nginx/static/public  && /usr/bin/docker cp dash:/app/amplitude/dash/public /mnt/nginx/static - Function: cmd.wait - Result: Clean
  Name: /mnt/nginx/conf.d/amplitude.conf - Function: file.managed - Result: Clean
  Name: /mnt/nginx/proxy_params/proxy_params - Function: file.managed - Result: Clean
  Name: /mnt/nginx/conf.d/blockips.conf - Function: file.managed - Result: Clean
  Name: /usr/bin/docker kill -s HUP nginx - Function: cmd.wait - Result: Clean
  Name: /etc/hosts - Function: file.managed - Result: Clean
  Name: /etc/security/limits.d/001-core-limits.conf - Function: file.managed - Result: Clean
  Name: /mnt - Function: mount.mounted - Result: Clean
  Name: ~admin/.ssh/authorized_keys - Function: file.managed - Result: Clean
  Name: deb http://apt.datadoghq.com/ stable main - Function: pkgrepo.managed - Result: Clean
  Name: datadog-agent - Function: pkg.latest - Result: Clean
  Name: /etc/dd-agent/datadog.conf - Function: file.managed - Result: Clean
  Name: datadog-agent - Function: service.running - Result: Clean
  Name: /usr/local/bin/dogcat - Function: file.managed - Result: Clean
  Name: /bin/systemctl daemon-reload - Function: cmd.wait - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh - Function: file.managed - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh -k "$lic" -x "$key" -y "$secret" -v "$ver" - Function: cmd.run - Result: Changed
  Name: /mnt/tmp/docker - Function: file.directory - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: file.managed - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: cmd.run - Result: Changed

Summary
-------------
Succeeded: 46 (changed=7)
Failed:     0
-------------
Total states run:     46
```
```
admin@saltshaker-001:~$
admin@saltshaker-001:~$ sudo amp_salt_deploy G@ec2_tags:cluster:dash '{"app_deploy": true}' 2

Executing run on ['dev-dash-002.us-west-2', 'prod-dash-005.us-west-2']

prod-dash-005.us-west-2:
  Name: core_packages - Function: pkg.installed - Result: Clean
  Name: deb https://get.docker.com/ubuntu docker main - Function: pkgrepo.managed - Result: Clean
  Name: lxc-docker - Function: pkg.latest - Result: Clean
  Name: /etc/default/docker - Function: file.managed - Result: Clean
  Name: docker - Function: service.running - Result: Clean
  Name: pip - Function: pip.installed - Result: Changed
  Name: requests - Function: pip.installed - Result: Changed
  Name: docker-py ==1.2.3 - Function: pip.installed - Result: Clean
  Name: ~admin/.ssh - Function: file.directory - Result: Clean
  Name: ~admin/.ssh/vault_id_rsa - Function: file.managed - Result: Clean
  Name: git@github.com:amplitude/vault.git - Function: git.latest - Result: Clean
  Name: /mnt/tmp/logs - Function: file.directory - Result: Clean
  Name: /mnt/tmp/logs/dash - Function: file.directory - Result: Clean
  Name: amplitudemobile/dash - Function: docker.pulled - Result: Clean
  Name: nginx - Function: docker.pulled - Result: Clean
  Name: /usr/local/bin/docker_compare_running - Function: file.managed - Result: Clean
  Name: /usr/bin/docker stop -t 10 dash && /usr/bin/docker rm -f dash - Function: cmd.run - Result: Clean
  Name: /usr/bin/docker stop -t 10 nginx && /usr/bin/docker rm -f nginx - Function: cmd.run - Result: Changed
  Name: dash - Function: docker.installed - Result: Clean
  Name: nginx - Function: docker.installed - Result: Changed
  Name: dash - Function: docker.running - Result: Clean
  Name: nginx - Function: docker.running - Result: Changed
  Name: /mnt/tmp/logs/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx - Function: file.directory - Result: Clean
  Name: /mnt/nginx/conf.d - Function: file.directory - Result: Clean
  Name: /mnt/nginx/static - Function: file.directory - Result: Clean
  Name: /bin/rm -rf /mnt/nginx/static/public  && /usr/bin/docker cp dash:/app/amplitude/dash/public /mnt/nginx/static - Function: cmd.wait - Result: Clean
  Name: /mnt/nginx/conf.d/amplitude.conf - Function: file.managed - Result: Clean
  Name: /mnt/nginx/proxy_params/proxy_params - Function: file.managed - Result: Clean
  Name: /mnt/nginx/conf.d/blockips.conf - Function: file.managed - Result: Clean
  Name: /usr/bin/docker kill -s HUP nginx - Function: cmd.wait - Result: Clean
  Name: /etc/hosts - Function: file.managed - Result: Clean
  Name: /etc/security/limits.d/001-core-limits.conf - Function: file.managed - Result: Clean
  Name: /mnt - Function: mount.mounted - Result: Clean
  Name: ~admin/.ssh/authorized_keys - Function: file.managed - Result: Clean
  Name: deb http://apt.datadoghq.com/ stable main - Function: pkgrepo.managed - Result: Clean
  Name: datadog-agent - Function: pkg.latest - Result: Clean
  Name: /etc/dd-agent/datadog.conf - Function: file.managed - Result: Clean
  Name: datadog-agent - Function: service.running - Result: Clean
  Name: /usr/local/bin/dogcat - Function: file.managed - Result: Clean
  Name: /bin/systemctl daemon-reload - Function: cmd.wait - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh - Function: file.managed - Result: Clean
  Name: /usr/local/bin/opsclarity-agent-installer.sh -k "$lic" -x "$key" -y "$secret" -v "$ver" - Function: cmd.run - Result: Changed
  Name: /mnt/tmp/docker - Function: file.directory - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: file.managed - Result: Clean
  Name: /usr/local/bin/dockerclean - Function: cmd.run - Result: Changed

Summary
-------------
Succeeded: 46 (changed=7)
Failed:     0
-------------
Total states run:     46

Executing run on ['prod-dash-006.us-west-2']

              ----------
              dev-dash-002.us-west-2:
                  ----------

Executing run on ['stag-dash-002.us-west-2']

              ----------
              prod-dash-006.us-west-2:
                  ----------

Executing run on ['prod-dash-003.us-west-2']

              ----------
              stag-dash-002.us-west-2:
                  ----------

Executing run on ['prod-dash-004.us-west-2']



              ----------
              prod-dash-003.us-west-2:
                  ----------
              ----------
              prod-dash-004.us-west-2:
                  ----------
```

amp_salt_deploy
```
search=$1
pillar_data=$2
batch_size=$3

if [ "x${batch_size}" == "x" ];then
  batch=''
else
  batch="-b ${batch_size}"
fi

if [ "x${pillar_data}" == "x" ];then
  pillar=''
else
  pillar="pillar='${pillar_data}'"
fi

salt_cmd="sudo /usr/bin/salt ${batch} -C '${search}' state.highstate ${pillar}"

eval "${salt_cmd}"
```

**Comment by jfindlay** [comment]
@nicholascapo, @jaybocc2, thanks for the report.

**Comment by ze42** [comment]
The `-b N` quickly becomes `-b 1` in the middle of a salt run, is by design.

Previously, salt -b X was launching X minions, waiting for **ALL** results before launching X minions again.

In more recent versions, it launches X minions, waits for some reply, and launches again some minions so X are running.

In other word, it could be:
  * it launches 5 minions
  * 2 minions finishes together, it launches 2 new minions, with the 3 old still running.
  * 3 minions finishes together, 2 still running, it launches 3 new minions if there are still enough targets

That part should theoricaly greatly help with a few not-respnoding minions blocking a batch mode completly.

**Comment by ze42** [comment]
But there are some minions that do NOT show their reply at all in current batch mode.

See #26254

**Comment by nicholascapo** [comment]
This 'rolling' behavior seems a little strange to me, but I suppose it fine (just unexpected).

I am also seeing the problem with missing return data from some minions.

This ticket can be closed in favor of #26254

**Comment by jfindlay** [comment]
Closing in favor of #26254.
